elfloader-tool/src/arch-riscv/crt0.S

```diff
@@ -62,6 +67,7 @@ hsm_switch_hart:
   li a2, 0       // passed in a1 when new hart starts, this is the logical hart_id
   ecall
 
+  mv a0, t3 // Copy the ID of the HART again, since we've clobbered a0
```

Comment:
How is t3 still guaranteed to have the hart id in it? Isn't t3 caller-save so can be clobbered by the ecall?

Reply:
Perhaps? I'm not well versed in RISC-V assembly to know that this is the case. From what I've tested on the Hifive, this is works fine even when the boot HART is not HART 1.

Reply:
See https://github.com/riscv/riscv-elf-psabi-doc/blob/master/riscv-elf.md#register-convention, should use s0 (x8) then instead of t3 (x28)

Reply:
Cool, I'll change that.